src/preact/component/intersection-observer-resize.js

```diff
@@ -0,0 +1,50 @@
+import {toWin} from '#core/window';
+
+import {useEffect} from '#preact';
+
+const ioForWindow = new WeakMap();
+const callbackMap = new WeakMap();
+
+/**
+ * Uses a shared IntersectionObserver per window instance to observe the given `ref`.
+ *
+ * @param {{current: Element}} ref
+ * @param {function(IntersectionObserverEntry)} callback
+ * @param {?Window} targetWin
+ */
+export function useIntersectionObserver(ref, callback, targetWin = null) {
+  useEffect(() => {
+    const node = ref.current;
+    if (!node) {
+      return;
+    }
+    // If a window instance is not provided, uses the one local to the given `ref`.
+    const win = targetWin ?? toWin(node.ownerDocument?.defaultView);
+    if (!win) {
+      return;
+    }
+    callbackMap.set(node, callback);
+
+    if (ioForWindow.has(win)) {
+      ioForWindow.get(win).observe(node);
+    } else {
+      const io = new win.IntersectionObserver((entries) => {
+        entries.reduceRight((accumulator, entry) => {
```

Comment:
Nit: IMO it would be easier to read if we didn't use reduce. Especially given that `reduce` is meant to be more about creating a result than performing side-effects.

```js
const seen = new Set();
for (let i = entries.length - 1; i >= 0; i--) {
  const {target} = entries[i];
  if (seen.has(target)) {
    continue;
  }
  seen.add(target);
  callbackMap.get(target)(entry);
}
```